# Day 12 - Deepseek R1 Lite

In [8]:
def main():
    import sys
    sys.setrecursionlimit(1000000)
    
    with open('input.txt', 'r') as f:
        grid = [list(line.strip()) for line in f.readlines()]
    
    rows = len(grid)
    if rows > 0:
        cols = len(grid[0])
    else:
        cols = 0
    
    visited = [[False for _ in range(cols)] for _ in range(rows)]
    
    def dfs(r, c, plant_type):
        area = 1
        perimeter = 0
        stack = [(r, c)]
        visited[r][c] = True
        while stack:
            x, y = stack.pop()
            for dx, dy in [(-1,0),(1,0),(0,-1),(0,1)]:
                nx, ny = x + dx, y + dy
                if 0 <= nx < rows and 0 <= ny < cols:
                    if grid[nx][ny] == plant_type and not visited[nx][ny]:
                        visited[nx][ny] = True
                        stack.append((nx, ny))
                        area += 1
                    elif grid[nx][ny] != plant_type:
                        perimeter += 1
                else:
                    perimeter += 1
        return area, perimeter
    
    total_price = 0
    for i in range(rows):
        for j in range(cols):
            if not visited[i][j]:
                plant_type = grid[i][j]
                area, perimeter = dfs(i, j, plant_type)
                price = area * perimeter
                total_price += price
    
    print(total_price)

if __name__ == "__main__":
    main()

1449902


## Part 2

In [10]:
def main():
    import sys
    sys.setrecursionlimit(1000000)
    
    with open('input.txt', 'r') as f:
        grid = [list(line.strip()) for line in f.readlines()]
    
    rows = len(grid)
    if rows > 0:
        cols = len(grid[0])
    else:
        cols = 0
    
    visited = [[False for _ in range(cols)] for _ in range(rows)]
    
    def dfs(r, c, plant_type):
        stack = [(r, c)]
        visited[r][c] = True
        fence_edges = []
        area = 1
        
        while stack:
            x, y = stack.pop()
            
            # Check top edge
            if x - 1 < 0 or grid[x-1][y] != plant_type:
                fence_edges.append((x, y, 'top'))
            # Check bottom edge
            if x + 1 >= rows or grid[x+1][y] != plant_type:
                fence_edges.append((x, y, 'bottom'))
            # Check left edge
            if y - 1 < 0 or grid[x][y-1] != plant_type:
                fence_edges.append((x, y, 'left'))
            # Check right edge
            if y + 1 >= cols or grid[x][y+1] != plant_type:
                fence_edges.append((x, y, 'right'))
            
            for dx, dy in [(-1,0),(1,0),(0,-1),(0,1)]:
                nx, ny = x + dx, y + dy
                if 0 <= nx < rows and 0 <= ny < cols:
                    if grid[nx][ny] == plant_type and not visited[nx][ny]:
                        visited[nx][ny] = True
                        stack.append((nx, ny))
                        area += 1
        
        # Function to perform BFS on fence edges
        def bfs(edges_set, start_edge, direction, visited_edges):
            from collections import deque
            queue = deque([start_edge])
            visited_edges.add(start_edge)
            while queue:
                edge = queue.popleft()
                if direction == 'horizontal':
                    r, c, d = edge
                    neighbors = []
                    if d == 'top':
                        neighbors.append((r, c-1, 'top'))
                        neighbors.append((r, c+1, 'top'))
                    elif d == 'bottom':
                        neighbors.append((r, c-1, 'bottom'))
                        neighbors.append((r, c+1, 'bottom'))
                    for neighbor in neighbors:
                        if neighbor in edges_set and neighbor not in visited_edges:
                            queue.append(neighbor)
                            visited_edges.add(neighbor)
                elif direction == 'vertical':
                    r, c, d = edge
                    neighbors = []
                    if d == 'left':
                        neighbors.append((r-1, c, 'left'))
                        neighbors.append((r+1, c, 'left'))
                    elif d == 'right':
                        neighbors.append((r-1, c, 'right'))
                        neighbors.append((r+1, c, 'right'))
                    for neighbor in neighbors:
                        if neighbor in edges_set and neighbor not in visited_edges:
                            queue.append(neighbor)
                            visited_edges.add(neighbor)
        
        # Separate horizontal and vertical edges
        horizontal_edges = set()
        vertical_edges = set()
        for edge in fence_edges:
            if edge[2] in ['top', 'bottom']:
                horizontal_edges.add(edge)
            else:
                vertical_edges.add(edge)
        
        # Count horizontal sides
        num_horizontal_sides = 0
        visited_horizontal = set()
        edges_to_process = horizontal_edges.copy()
        while edges_to_process:
            edge = edges_to_process.pop()
            if edge not in visited_horizontal:
                bfs(horizontal_edges, edge, 'horizontal', visited_horizontal)
                num_horizontal_sides += 1
        
        # Count vertical sides
        num_vertical_sides = 0
        visited_vertical = set()
        edges_to_process = vertical_edges.copy()
        while edges_to_process:
            edge = edges_to_process.pop()
            if edge not in visited_vertical:
                bfs(vertical_edges, edge, 'vertical', visited_vertical)
                num_vertical_sides += 1
        
        num_sides = num_horizontal_sides + num_vertical_sides
        price = area * num_sides
        return price
    
    total_price = 0
    for i in range(rows):
        for j in range(cols):
            if not visited[i][j]:
                plant_type = grid[i][j]
                price = dfs(i, j, plant_type)
                total_price += price
    
    print(total_price)

if __name__ == "__main__":
    main()

908042


1405439 - too high
1171983 - too high
1086038 - too high
937813  - too high
908042  - correct
897612  - too low
893165  - too low
857063  - too low
66401   - too low
39053   - too low
35764   - too low